Support correct specification of project descriptor

jjbustamante opened the pull request:
## Problem

The CNB project descriptor v0.2 was [updated](https://github.com/buildpacks/spec/blob/main/extensions/project-descriptor.md#iobuildpacksbuildenv-optional) and the way the environment variables are set was changed. We need to add support for this.

## Solution

A new structured was added for parsing the new way of setting environment variables in the project descriptor, but we also kept the previous one for backward compatibility. 

Fixes #1017 

Signed-off-by: Juan Bustamante <jbustamante@vmware.com>

samj1912 commented on pkg/cnb/project_descriptor_test.go:
```diff
@@ -243,12 +244,41 @@ value = "valueC"
-						checkEnvVar(t, platformDir, "keyB", "valueB")
-						checkEnvVar(t, platformDir, "keyC", "valueAnotherC")
+
+					when("vars where set with new v0.2 project.toml", func() {
```
@jjbustamante can we also add a test case where both are set? The new one should have higher precedence

jjbustamante replied:
Interesting.

Yes, let me add that case, thanks @samj1912 !

tylerphelan commented on pkg/cnb/project_descriptor.go:
```diff
@@ -105,6 +107,35 @@ func processFiles(appDir string, d build) error {
+	return env
+}
+
+func mergeEnvVariables(old []envVariable, new []envVariable) []envVariable {
```
pack appears to just use the newer field instead of merge: https://github.com/buildpacks/pack/pull/1479/files#diff-4216f3c5d147f7361545afe9fb76c1fe23790f314e3f39ae1eca8559d726bd09R52

maybe this can help simplify things

anyone have opinions? @tomkennedy513 @samj1912 @matthewmcnew @ncarlson @yaelharel @chenbh

tomkennedy513 replied:
Ya, i think if the newer field is populated just use that

jjbustamante replied:
Do you mean to keep it how it was before?

tylerphelan replied:
I think it was backwards before:

pack:
```
env := versionedDescriptor.IO.Buildpacks.Build.Env
if env == nil {
    env = versionedDescriptor.IO.Buildpacks.Env.Build
}
```

before:

```
env := d.IO.Buildpacks.Env
if env == nil {
	env = d.IO.Buildpacks.BuildEnv.Env
}
```

jjbustamante replied:
Yeap, I added the logic before following @samj1912 [comment](https://github.com/pivotal/kpack/pull/1027#discussion_r961150010)

jjbustamante commented:
I just added more unit testing to the merge logic. I agreed that pack has not merging logic. If we agree to avoid that one I can revert to my first commit.

@samj1912, @tylerphelan

samj1912 commented:
@jjbustamante what I meant by https://github.com/pivotal/kpack/pull/1027#discussion_r961150010 was that if both the fields are set, the new field should have higher precedence and we should ignore the old one. Apologies if it came across as merging the two. We should also have a test case similar to pack where we check that the env vars are from the new field and not the old one.

jjbustamante commented:
> @jjbustamante what I meant by [#1027 (comment)](https://github.com/pivotal/kpack/pull/1027#discussion_r961150010) was that if both the fields are set, the new field should have higher precedence and we should ignore the old one. Apologies if it came across as merging the two. We should also have a test case similar to pack where we check that the env vars are from the new field and not the old one.

Thanks @samj1912 for the clarification, I removed the merging code and the test case added was also modified. If you defined environment variables in the old version and the on the new version, kpack is going to take ONLY the one defines in the new version.

tomkennedy513 commented on pkg/cnb/project_descriptor.go:
```diff
@@ -105,6 +107,22 @@ func processFiles(appDir string, d build) error {
+	if env == nil {
+		env = d.IO.Buildpacks.BuildEnv.Env
+	} else {
+		if d.IO.Buildpacks.BuildEnv.Env != nil {
```
Actually my bad, i misread which was the newer key, but couldnt you set teh env to the newer key and if the env is nil then set it to the older key

tylerphelan replied:
does this work?

```
func getEnvVariables(d descriptorV2) []envVariable {
  env := d.IO.Buildpacks.BuildEnv.Env
  if env == nil {
    env = d.IO.Buildpacks.Env
  }
  return env
}
```

jjbustamante replied:
Yeah, I realized that after pushing hahaha

samj1912 commented on pkg/cnb/project_descriptor_test.go:
```diff
@@ -243,12 +244,61 @@ value = "valueC"
+							ioutil.WriteFile(projectToml, []byte(`
+[_]
+schema-version = "0.2"
+[[io.buildpacks.env]]
```
@jjbustamante shouldn't this be io.buildpacks.env.build?

jjbustamante replied:
That one is defined in L293 see [here](https://github.com/pivotal/kpack/blob/0e3c15f9d5230ff454dbc1776362e7a3f6c7a966/pkg/cnb/project_descriptor_test.go#L289-L295)

In that case, I defined the same variable twice:

```toml
schema-version = "0.2"
[[io.buildpacks.env]]
name = "keyA"
value = "valueA"
# check that new declaration must have higher precedence
[[io.buildpacks.build.env]]
name = "keyA"
value = "newValueA"
```

And then in the assert, I verify the one taken was `newValueA`

samj1912 replied:
@jjbustamante I meant to comment on the key itself. Left some notes at https://github.com/pivotal/kpack/pull/1027#pullrequestreview-1095151602

samj1912 commented on pkg/cnb/project_descriptor_test.go:
```diff
@@ -243,12 +244,61 @@ value = "valueC"
+							ioutil.WriteFile(projectToml, []byte(`
+[_]
+schema-version = "0.2"
+[[io.buildpacks.env]]
```
```suggestion
[[io.buildpacks.env.build]]
```

samj1912 reviewed (changes requested):
The current 'old' code in kpack seems a bit incorrect? See output section of https://github.com/buildpacks/pack/pull/1479

for the old key vs the new key

jjbustamante commented:
> The current 'old' code in kpack seems a bit incorrect? See output section of [buildpacks/pack#1479](https://github.com/buildpacks/pack/pull/1479)
> 
> for the old key vs the new key
> 
> old pack key - io.buildpacks.env.build
> 
> new pack key - io.buildpacks.build.env
> 
> old kpack key - io.buildpacks.env? (this seems incorrect?)

Oh! That's interesting. @matthewmcnew @tylerphelan @tomkennedy513  any thoughts on that? I didn't touch that part

tylerphelan commented on pkg/cnb/project_descriptor.go:
```diff
@@ -105,6 +107,17 @@ func processFiles(appDir string, d build) error {
+func getEnvVariables(d descriptorV2) []envVariable {
+	env := d.IO.Buildpacks.BuildEnv.Env
+	if env == nil {
+		env = d.IO.Buildpacks.Env
```
let's start the a thread here per: https://github.com/pivotal/kpack/pull/1027#issuecomment-1235771433

> The current 'old' code in kpack seems a bit incorrect? See output section of https://github.com/buildpacks/pack/pull/1479

It appears we incorrectly used ` io.buildpacks.env` while `pack` incorrectly used `io.buildpacks.env.build`.

We can either support all three or ignore support for `io.buildpacks.env.build`

@samj1912 @tomkennedy513 @matthewmcnew @jjbustamante 
for the old key vs the new key

old pack key - io.buildpacks.env.build

new pack key - io.buildpacks.build.env

old kpack key - io.buildpacks.env? (this seems incorrect?)

matthewmcnew replied:
Kpack exists in the pack ecosystem where users would reasonably expect kpack to behave similarly to pack. So, I think we should provide backwards compatibility to both `io.buildpacks.env.build` and kpack incorrect key: `io.buildpacks.env`.

matthewmcnew replied:
Good 👀 @samj1912 !

samj1912 commented:
So IIUC

highest to lowest precedence -

- io.buildpacks.build.env
- io.buildpacks.env.build
- io.buildpacks.env
?

jjbustamante commented:
> * io.buildpacks.build.env
> * io.buildpacks.env.build
> * io.buildpacks.env

I agree with the major reason to do it like this, but, if I am not wrong, I am not able to use the same *structs* to deal with the 3 cases. The struct to handle `build.env` and `env.build` can be the same, with two fields but if I get an error trying to parse the toml file then I will have to create a different *struct* to try to parse the case `env`, right?

jjbustamante commented:
I am not totally happy, but I did a refactor to include the compatibility for all the 3 cases:

- io.buildpacks.build.env
- io.buildpacks.env.build
- io.buildpacks.env

If anyone have suggestions about the approach, let me know.

matthewmcnew commented on pkg/cnb/project_descriptor.go:
```diff
@@ -41,13 +41,22 @@ func ProcessProjectDescriptor(appDir, descriptorPath, platformDir string, logger
 	if _, err := toml.DecodeFile(file, &d); err != nil {
-		return descriptorV2{}, err
+		var dV2NoSpec descriptorV2NoSpec
+		if _, err2 := toml.DecodeFile(file, &dV2NoSpec); err2 != nil {
```
Would it be easier to use the `encoding.TextUnmarshaler` interface to add logic to unmarshall the different types of `env` fields to avoid a whole different type?

jjbustamante replied:
@matthewmcnew you mean to use [this](https://github.com/BurntSushi/toml#using-the-marshaler-and-encodingtextunmarshaler-interfaces)?

matthewmcnew replied:
Yeah!

jjbustamante replied:
Thanks for the help @matthewmcnew. Done

matthewmcnew commented on pkg/cnb/project_descriptor.go:
```diff
@@ -105,6 +119,17 @@ func processFiles(appDir string, d build) error {
+// Because CNB Project Descriptor v0.2 has two ways for defining environment variables.
+// see https://github.com/buildpacks/spec/blob/main/extensions/project-descriptor.md#iobuildpacksbuildenv-optional
+// This function calculates the final environment variables
+func getEnvVariables(d descriptorV2) []envVariable {
```
Can we add this instead as a method on `descriptorV2` so that consumers of  descriptorV2 don't need to care about the heirachy of env and can just ask ` descriptor.Env()`?

jjbustamante replied:
Done

tylerphelan commented on pkg/cnb/env_vars.go:
```diff
@@ -27,3 +28,37 @@ type envVariable struct {
+	Env []envVariable
+}
+
+func (a *envBuildVariable) UnmarshalTOML(f interface{}) error {
```
this will prioritize `io.buildpacks.build` over legacy pack's `io.buildpacks.env.build` but I think it's ok

jjbustamante replied:
In fact, I think it is not giving any priority, it is just parsing both cases depending on what the user specifies because they go to different **case** in the switch.

tylerphelan replied:
yea I think this is ok b/c `io.buildpacks.env` and `io.buildpacks.env.build` is not valid toml

matthewmcnew commented on pkg/cnb/env_vars.go:
```diff
@@ -27,3 +28,37 @@ type envVariable struct {
+	var e []envVariable
+	for _, env := range v {
+		e = append(e, envVariable{
+			Name:  env["name"].(string),
```
So, I think we need to add the valid cast check.

matthewmcnew replied:
Sorry, I dropped that for brevity earlier.

jjbustamante replied:
Well, if I add a test case with the following configuration

``` toml
schema-version = "0.2"
[[io.buildpacks.env.build]]
name = keyA
value = "valueA"
```
It actually fails with the following message: "toml: line 5 (last key \"io.buildpacks.env.build.name\"): expected value but found \"keyA\" instead" 

It is actually not reaching the **UnmarshalTOML** implementations, it fails [here](https://github.com/BurntSushi/toml/blob/891d2617ddbdfa265c4095b53103c010c98e6591/parse.go#L98)

tylerphelan replied:
try with an int

jjbustamante replied:
With an **int** it actually gets into the code we want to protect, I just added 4 more test cases and some validations

tomkennedy513 commented:
@jjbustamante can you rebase this pr since its out of date

jjbustamante commented:
> @jjbustamante can you rebase this pr since its out of date

Done @tomkennedy513
I also squashed the commits.
